python/lsst/dm/OCPS/ocps_csc.py

```diff
@@ -0,0 +1,230 @@
+
+    This CSC executes pipelines on specified data.
+
+    Parameters
+    ----------
+    simulation_mode: `int` (optional)
+        Simulation mode.
+
+    Raises
+    ------
+    salobj.ExpectedError
+        If simulation_mode is invalid.
+
+    Notes
+    -----
+    **Simulation Modes**
+
+    Supported simulation modes:
+
+    * 0: regular operation
+    * 1: simulation mode:
+      * true.yaml always succeeds
+      * false.yaml always fails
+      * error.yaml always gives a connection error
+
+    ** Error Codes**
+
+    * 1: could not connect to the execution service endpoint
+    * 2: pipeline failed in synchronous mode
+    """
+
+    valid_simulation_modes = (0, 1)
+    version = __version__
+
+    def __init__(
+        self, config_dir=None, initial_state=salobj.State.STANDBY,
+        settings_to_apply='', simulation_mode=0,
+    ):
+        self.config = None
+        self.simulated_jobs = set()
+        super().__init__(
+            "OCPS",
+            index=0,
+            config_schema=yaml.safe_load(CONFIG_SCHEMA),
+            config_dir=config_dir,
+            initial_state=initial_state,
+            settings_to_apply=settings_to_apply,
+            simulation_mode=simulation_mode,
+        )
+        self.cmd_execute.allow_multiple_callbacks = True
+
+    async def do_execute(self, data):
+        """Implement the ``execute`` command.
+
+        Submit a request for execution to the back-end REST API.
+
+        Parameters
+        ----------
+        data: types.SimpleNamespace
+            Must contain pipeline, version, and config attributes
+        """
+        self.assert_enabled("execute")
+        self.log.info(f"execute command with {data}")
+
+        if self.simulation_mode == 0:
+            payload = dict(
+                type="pipeline",
+                version=data.version,
+                config=data.config,
+                env=[dict(name="PIPELINE", value=data.pipeline)],
+            )
+            self.log.info(f"PUT: {payload}")
+            result = self.connection.put(self.config.url, json=payload)
+            result.raise_for_status()
+            self.log.info(f"PUT result: {result.data}")
+            job_id = result.json().job_id
+            status_url = f"{self.config.url}/{job_id}"
+        else:
+            if data.pipeline not in ("true.yaml", "false.yaml", "fault.yaml"):
+                raise salobj.ExpectedError(f"Unknown pipeline: {data.pipeline}")
+            job_id = f"{data.pipeline}-{time.time()}"
+            status_url = f"ocps://{job_id}"
+            self.log.info(f"Simulated POST result: {status_url}")
+            self.simulated_jobs.add(job_id)
+
+        payload = json.dumps(dict(job_id=job_id))
+        self.cmd_execute.ack_in_progress(data, timeout=600.0, result=payload)
+        self.log.info(f"Ack in progress: {payload}")
+        self.log.info(f"Starting async wait: {status_url}")
+
+        while True:
+            if status_url.startswith("ocps://"):
```

Comment:
I'm confused. You say "simulation mode" 2 lines below, and this branch clearly has something to do with simulation (you could surely not wait for a random long time otherwise). But this branch appears to be orthogonal to simulation mode -- the user can configure the URL to be ocps or not, independently of simulation mode.

Is that an important feature? Because if not, I think the code and behavior would be much easier to understand if this branch run if and only if simulation mode = 1 and you eliminated the special oscp:// URL prefix.

Reply:
In simulation mode, the configured URL is ignored and this special `ocps://` URL is used internally to signal the mode to the REST API section.  It isn't intended that an `ocps://` URL ever be configured explicitly.  I will document this better.